io-sim-classes/src/Control/Monad/Class/MonadTimer.hs

```diff
@@ -178,16 +178,26 @@ instance MonadTimer IO where
 
   timeout = IO.timeout . diffTimeToMicrosecondsAsInt
 
+threadDelayIO :: Integer -> IO ()
+threadDelayIO time = do
+    IO.threadDelay $ fromInteger maxDelay
+    when (maxDelay /= time) $ threadDelayIO (time - maxDelay)
+  where
+    maxDelay = min time $ toInteger (maxBound :: Int)
 
 diffTimeToMicrosecondsAsInt :: DiffTime -> Int
 diffTimeToMicrosecondsAsInt d =
     let usec :: Integer
-        usec = diffTimeToPicoseconds d `div` 1000000 in
+        usec = diffTimeToMicrosecondsAsInteger d in
     -- Can only represent usec times that fit within an Int, which on 32bit
     -- systems means 2^31 usec, which is only ~35 minutes.
     assert (usec <= fromIntegral (maxBound :: Int)) $
     fromIntegral usec
```

Comment:
This function is the cause of the problem: the assert can fail.

You address it for `threadDelay`, but the problematic function is used by other functions as well.

I don't think @dcoutts (who wrote `io-sim(-classes)`) would be satisfied that we just fix one function as opposed to fixing it properly once and for all.

However, I think we, consensus, can work around the issue entirely on our side.

Plan:
* Undo the changes to `MonadTimer`
* Add a function to consensus:
  ```haskell
  -- | Document that this is for sleeping longer periods, mention the overflow and ~35 minute limit
  threadSleep :: MonadTimer m => DiffTime -> m ()
  threadSleep = <do the loop>
  ```
  (I first had a new `MonadSleep` class in mind, but a function is actually enough).
* Look for all uses of `threadDelay` in consensus and replace them with `threadSleep` where the chance exists that we might have to wait longer than 35 minutes. In particular, the place where we wait for the blockchain time to start.

Reply:
Wouldn't it better to leave the changes in MonadTImer? This way all users of the class will have the fix, without needing any wrapper.

This function is only used on Timeout, other than threadDelay. I could take a look if this is also fixable, if we want to entirely remove it.

Reply:
> Wouldn't it better to leave the changes in MonadTImer? This way all users of the class will have the fix, without needing any wrapper.

Sure, that would be ideal. But I'm not sure how easy that is.

> This function is only used on Timeout, other than threadDelay. I could take a look if this is also fixable, if we want to entirely remove it.

In that case you'll have to ask a review from @dcoutts.


First do the `threadSleep` fix, which should be merged ASAP, as people launching testnets that start in more than 35 minutes will run into the bug.

Reply:
I replaced threadDelay with threadSleep in some places, hopefully I didn't miss something important.

Also I assumed your quotes and replies are reversed at your previous response?